// @gnosis.pm/zodiac/dist/hardhat.config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require("@nomiclabs/hardhat-etherscan");
require("@nomiclabs/hardhat-waffle");
require("solidity-coverage");
require("hardhat-deploy");
const dotenv_1 = __importDefault(require("dotenv"));
const yargs_1 = __importDefault(require("yargs"));
const { network } = yargs_1.default
    .option("network", {
    type: "string",
    default: "hardhat",
})
    .help(false)
    .version(false)
    .parseSync();
// Load environment variables.
dotenv_1.default.config();
const { INFURA_KEY, MNEMONIC, ETHERSCAN_API_KEY, PK } = process.env;
require("./src/tasks/singleton-deployment");
const DEFAULT_MNEMONIC = "candy maple cake sugar pudding cream honey rich smooth crumble sweet treat";
const sharedNetworkConfig = {};
if (PK) {
    sharedNetworkConfig.accounts = [PK];
}
else {
    sharedNetworkConfig.accounts = {
        mnemonic: MNEMONIC || DEFAULT_MNEMONIC,
    };
}
if (["mainnet", "goerli", "ropsten"].includes(network) &&
    INFURA_KEY === undefined) {
    throw new Error(`Could not find Infura key in env, unable to connect to network ${network}`);
}
exports.default = {
    paths: {
        artifacts: "build/artifacts",
        cache: "build/cache",
        deploy: "src/deploy",
        sources: "contracts",
    },
    solidity: {
        compilers: [{ version: "0.8.6" }, { version: "0.6.12" }],
    },
    networks: {
        mainnet: {
            ...sharedNetworkConfig,
            url: `https://mainnet.infura.io/v3/${INFURA_KEY}`,
        },
        ropsten: {
            ...sharedNetworkConfig,
            url: `https://ropsten.infura.io/v3/${INFURA_KEY}`,
        },
        goerli: {
            ...sharedNetworkConfig,
            url: `https://goerli.infura.io/v3/${INFURA_KEY}`,
        },
        gnosis: {
            ...sharedNetworkConfig,
            url: "https://rpc.gnosischain.com",
        },
        polygon: {
            ...sharedNetworkConfig,
            url: "https://rpc.ankr.com/polygon",
        },
        bsc: {
            ...sharedNetworkConfig,
            url: "https://bsc-dataseed.binance.org",
        },
        arbitrum: {
            ...sharedNetworkConfig,
            url: "https://arb1.arbitrum.io/rpc",
        },
        optimism: {
            ...sharedNetworkConfig,
            url: "https://mainnet.optimism.io",
        },
        avalanche: {
            ...sharedNetworkConfig,
            url: "https://rpc.ankr.com/avalanche",
        },
    },
    namedAccounts: {
        deployer: 0,
    },
    mocha: {
        timeout: 2000000,
    },
    etherscan: {
        apiKey: ETHERSCAN_API_KEY,
    },
};
