// @boardroom/gov-lib/dist/index.js
"use strict";
/*

  For libraries, its important to have intentional exports. Anything we
  export may end up being depended on by somebody, so we don't want to end up
  changing this without a reason. This is also the reason we have a snapshot
  test for the exported symbols of gov-sdk

  https://www.hyrumslaw.com

*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaggedTypeResolver = exports.bufferAsyncIterable = exports.iteratePages = exports.extractPage = exports.encodeCursor = exports.decodeCursor = exports.ProtocolRegistry = exports.createRegistry = exports.Protocol = void 0;
var protocol_1 = require("./protocol");
Object.defineProperty(exports, "Protocol", { enumerable: true, get: function () { return protocol_1.Protocol; } });
var framework_1 = require("./framework");
Object.defineProperty(exports, "createRegistry", { enumerable: true, get: function () { return framework_1.createRegistry; } });
var registry_1 = require("./registry");
Object.defineProperty(exports, "ProtocolRegistry", { enumerable: true, get: function () { return registry_1.ProtocolRegistry; } });
var pagination_1 = require("./pagination");
Object.defineProperty(exports, "decodeCursor", { enumerable: true, get: function () { return pagination_1.decodeCursor; } });
Object.defineProperty(exports, "encodeCursor", { enumerable: true, get: function () { return pagination_1.encodeCursor; } });
Object.defineProperty(exports, "extractPage", { enumerable: true, get: function () { return pagination_1.extractPage; } });
Object.defineProperty(exports, "iteratePages", { enumerable: true, get: function () { return pagination_1.iteratePages; } });
var iterables_1 = require("./iterables");
Object.defineProperty(exports, "bufferAsyncIterable", { enumerable: true, get: function () { return iterables_1.bufferAsyncIterable; } });
__exportStar(require("./errors"), exports);
var resolver_1 = require("./resolver");
Object.defineProperty(exports, "TaggedTypeResolver", { enumerable: true, get: function () { return resolver_1.TaggedTypeResolver; } });
// export everything from adapters -- watch that API surface area!
__exportStar(require("./adapters"), exports);
__exportStar(require("./adapters/proposals"), exports);
__exportStar(require("./adapters/treasury"), exports);
__exportStar(require("./adapters/token-info"), exports);
__exportStar(require("./adapters/icon"), exports);
__exportStar(require("./adapters/vote"), exports);
__exportStar(require("./adapters/vote-power"), exports);
__exportStar(require("./adapters/delegation"), exports);
__exportStar(require("./adapters/create-proposal"), exports);
__exportStar(require("./adapters/create-on-chain-proposal"), exports);
__exportStar(require("./adapters/proposal-execution"), exports);
__exportStar(require("./adapters/pod"), exports);
__exportStar(require("./adapters/common"), exports);
__exportStar(require("./adapters/general"), exports);
__exportStar(require("./adapters/staking-token"), exports);
// export everything from transports -- watch that API surface area!
__exportStar(require("./transports"), exports);
__exportStar(require("./transports/http"), exports);
__exportStar(require("./transports/ipfs"), exports);
__exportStar(require("./transports/network-transport"), exports);
__exportStar(require("./transports/signer"), exports);
