// @gnosis.pm/safe-core-sdk/dist/typechain/factories/IProxyCreationCallback__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IProxyCreationCallback__factory = void 0;
const ethers_1 = require("ethers");
class IProxyCreationCallback__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IProxyCreationCallback__factory = IProxyCreationCallback__factory;
const _abi = [
    {
        constant: false,
        inputs: [
            {
                internalType: "contract GnosisSafeProxy",
                name: "proxy",
                type: "address",
            },
            {
                internalType: "address",
                name: "_mastercopy",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "initializer",
                type: "bytes",
            },
            {
                internalType: "uint256",
                name: "saltNonce",
                type: "uint256",
            },
        ],
        name: "proxyCreated",
        outputs: [],
        payable: false,
        stateMutability: "nonpayable",
        type: "function",
    },
];
