// @gnosis.pm/safe-core-sdk/dist/src/utils/transactions/gas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.estimateGasForTransactionExecution = exports.estimateTxGas = void 0;
const safe_core_sdk_types_1 = require("@gnosis.pm/safe-core-sdk-types");
const ethers_1 = require("ethers");
function estimateDataGasCosts(data) {
    const reducer = (accumulator, currentValue) => {
        if (currentValue === '0x') {
            return accumulator + 0;
        }
        if (currentValue === '00') {
            return accumulator + 4;
        }
        return accumulator + 16;
    };
    return data.match(/.{2}/g).reduce(reducer, 0);
}
async function estimateTxGas(safeContract, to, valueInWei, data, operation) {
    let txGasEstimation = 0;
    const safeAddress = safeContract.address;
    const estimateData = safeContract.interface.encodeFunctionData('requiredTxGas', [
        to,
        valueInWei,
        data,
        operation
    ]);
    try {
        const estimateResponse = await safeContract.provider.call({
            to: safeAddress,
            from: safeAddress,
            data: estimateData
        }, 'latest');
        txGasEstimation = ethers_1.BigNumber.from('0x' + estimateResponse.substring(138)).toNumber() + 10000;
    }
    catch (error) { }
    if (txGasEstimation > 0) {
        const dataGasEstimation = estimateDataGasCosts(estimateData);
        let additionalGas = 10000;
        for (let i = 0; i < 10; i++) {
            try {
                const estimateResponse = await safeContract.provider.call({
                    to: safeAddress,
                    from: safeAddress,
                    data: estimateData,
                    gasPrice: 0,
                    gasLimit: txGasEstimation + dataGasEstimation + additionalGas
                });
                if (estimateResponse !== '0x') {
                    break;
                }
            }
            catch (error) { }
            txGasEstimation += additionalGas;
            additionalGas *= 2;
        }
        return txGasEstimation + additionalGas;
    }
    try {
        const estimateGas = await safeContract.provider.estimateGas({
            to,
            from: safeAddress,
            value: valueInWei,
            data
        });
        return estimateGas.toNumber();
    }
    catch (error) {
        if (operation === safe_core_sdk_types_1.OperationType.DelegateCall) {
            return 0;
        }
        return Promise.reject(error);
    }
}
exports.estimateTxGas = estimateTxGas;
async function estimateGasForTransactionExecution(safeContract, from, tx) {
    try {
        const gas = await safeContract.estimateGas.execTransaction(tx.data.to, tx.data.value, tx.data.data, tx.data.operation, tx.data.safeTxGas, tx.data.baseGas, tx.data.gasPrice, tx.data.gasToken, tx.data.refundReceiver, tx.encodedSignatures(), { from, gasPrice: tx.data.gasPrice });
        return gas.toNumber();
    }
    catch (error) {
        return Promise.reject(error);
    }
}
exports.estimateGasForTransactionExecution = estimateGasForTransactionExecution;
