// @gnosis.pm/zodiac/dist/src/factory/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnownContracts = void 0;
var KnownContracts;
(function (KnownContracts) {
    KnownContracts["META_GUARD"] = "metaGuard";
    KnownContracts["OPTIMISTIC_GOVERNOR"] = "optimisticGovernor";
    KnownContracts["TELLOR"] = "tellor";
    KnownContracts["REALITY_ETH"] = "realityETH";
    KnownContracts["REALITY_ERC20"] = "realityERC20";
    KnownContracts["BRIDGE"] = "bridge";
    KnownContracts["DELAY"] = "delay";
    KnownContracts["EXIT_ERC20"] = "exit";
    KnownContracts["EXIT_ERC721"] = "exitERC721";
    KnownContracts["CIRCULATING_SUPPLY_ERC20"] = "circulatingSupplyERC20";
    KnownContracts["CIRCULATING_SUPPLY_ERC721"] = "circulatingSupplyERC721";
    KnownContracts["SCOPE_GUARD"] = "scopeGuard";
    KnownContracts["FACTORY"] = "factory";
    KnownContracts["ROLES"] = "roles";
    KnownContracts["OZ_GOVERNOR"] = "ozGovernor";
    KnownContracts["ERC20_VOTES"] = "erc20Votes";
    KnownContracts["ERC721_VOTES"] = "erc721Votes";
})(KnownContracts = exports.KnownContracts || (exports.KnownContracts = {}));
